Update README.md

barlehmann opened the pull request:
This is a pull request by Bar Lehmann

barlehmann committed:
Update README.md

leej3 commented:
Looks good. My pointers as you embark on this is:

1) Keep the goal in mind. That goal will be to  tick the boxes of the rubric, do a fun/useful project, or a bit of both. The scope of the project is broad and may take a lot of code to implement. Don't get lost in the details of the code. Most of your effort from the perspective of grading is to create a repo of code that is well documented, well organized, and well tested. From the perspective of a scientist most of your effort should be spent learning how tools like MNE think about and implement the  functionality that you want. That means many hours of learning from their documentation rather than writing your own code. Although as the saying goes... a week of coding can save you an hour of thought.

2) Having a clear answer to point one will help you pitch the project. For example, you could acknowledge that you are duplicating (or wrapping) functionality of MNE for the purposes of learning. That is a lot safer than attempting to make a genuine contribution by creating a new tool. Generally the answer to such an endevour is "but have you seen the pre-existing tool X?". The most effective use of your time is to contribute to open science community-driven efforts in your field rather than working on your own to tool, putting it on GitHub, and calling it open.

3) Dropping the scope of the project might help you. Preprocessing and visualization  of this data require large amounts of expertise. You don't necessarily want to expose a user of your package (read grader of your project) to that expertise. If you are putting a lot of time into this and want to achieve all of these things you a more than welcome to and we can provide help as you encounter the various issues you run into. But at the very least define clear stepping stones along your path to glory so that if it all goes pear-shaped you have a clear achievement and a gradable project rather than "hmmm, it didn't quite come together". You show evidence of this, I just want to emphasize it.

barlehmann committed:
Add files via upload

Commiting first copy of pycharm (.py) python code file that I have been working on during the past month to read EEG files and begin preliminary exploration.

barlehmann committed:
Adding a table tabulating subjective reports of both mood, focus, along with date and time information, to compare with values from biometric insturments such as EEG and GSR. Also some initial coding in the main python file for reading and visualizing this information.

barlehmann committed:
adding edits of table

barlehmann committed:
Adding code for user to create customized plot and obtain customized pearson r for variables that would like to explore in Mood_Focus_Table from Excel

barlehmann committed:
renamed file folder

barlehmann committed:
moving project files into correct folder

barlehmann committed:
adding both the graph_function which is the function I hope to get some feedback on related to how to graph the input from a specific sensor. I am also adding the EEG file that is being analysed

barlehmann committed:
adding the EEG file in eyes-closed condition called AusEC.edf

barlehmann committed:
adding two separate functions, the Pearson_and_Scatterplt function which can do pearson correlation analysis on two columns of data such as mood and alpha amplitude, and also adding the append_table_function which can append this same overarching spreadsheet of subjective and biometric data from new records, updating it with new relevant entries/recordings.

barlehmann committed:
adding edits and updates to three separate function files, along with more concise and clear comments to guide reader through each function

barlehmann committed:
adding comments for pearson and scatterplot function

barlehmann committed:
adding further comments to pearson and scatterplot function to ensure good readability by user

barlehmann committed:
Deleting main file since this is not useful now

barlehmann committed:
not useful now since making function .py files

barlehmann committed:
not needed

barlehmann committed:
not needed

barlehmann committed:
not needed

barlehmann committed:
moving function files to tests folder

barlehmann committed:
adding new functions to test folder

barlehmann committed:
q
q
Merge branch 'master' of https://github.com/barlehmann/EEG_GSR_Wellness_Tracker

i
'bad files'

q

barlehmann committed:
deleting files that have moved to test folder

barlehmann committed:
Delete pdbrc.py

barlehmann committed:
Delete 2pdbrc.py

barlehmann committed:
updating projects deleted files from github

barlehmann committed:
created a setup file and edited readme file

barlehmann committed:
changing file locations and uploading new file of graphing functions

barlehmann committed:
moving files from tests folder back into this folder and adding an updated functions file as well

barlehmann committed:
this file should be in other folder and has already been moved to this main other folder

barlehmann committed:
this file should be in other folder and has already been moved to this main other folder

barlehmann committed:
this file should be in other folder and has already been moved to this main other folder

barlehmann committed:
adding setup file

barlehmann committed:
updating fileMerge branch 'master' of https://github.com/barlehmann/EEG_GSR_Wellness_Tracker

leej3 commented:
Looks good. I would just call the module graphing. "_functions" is a a little superfluous. You need a subdirectory for all of your python modules.

The functions make it much  more readable and interprettable. Good job. Checkout "docstrings": A lot of your comments would be appropriate content for those. In general [pep8](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=2ahUKEwjAjYqmxIjpAhU3hXIEHdIkAY4QFjAAegQIARAB&url=https%3A%2F%2Fwww.python.org%2Fdev%2Fpeps%2Fpep-0008%2F&usg=AOvVaw0xCyNbAyoecmbLoILApdrP) should give you a good idea of how to make your python code tidier and more standard (comments above function declarations is not especially standard.

Based on your [circleci failure](https://app.circleci.com/pipelines/github/biof309/project_spring_2020/141/workflows/c1ee4454-e151-4a5c-8feb-11db38d13566/jobs/141) you have included a double quote that is not the character (copy and pasting or something like that).

barlehmann committed:
adding updated versions of each of three files that were put into function format with changes related to docstrings as well

barlehmann committed:
already pushed updated version of this file

barlehmann committed:
already pushed updated version of this file

barlehmann committed:
already pushed updated version of this file

barlehmann committed:
adding new subdirectory

barlehmann committed:
adding new subdirectory for relevant functions / modules

barlehmann committed:
Delete eeg_gsr_wellness_tracker_functions

barlehmann commented:
@leej3, (or @emilyyaklich and @suhwanplee) just wanted to respond to your last message: 

> Looks good. I would just call the module graphing. "_functions" is a a little superfluous. You need a subdirectory for all of your python modules.
> The functions make it much more readable and interprettable. Good job. Checkout "docstrings": A lot of your comments would be appropriate content for those. In general pep8 should give you a good idea of how to make your python code tidier and more standard (comments above function declarations is not especially standard.
> Based on your circleci failure you have included a double quote that is not the character (copy and pasting or something like that)."

Thank you for the feedback, I have changed the names of the functions, and have also put them into the general project folder (and also doing some research on a way to create the subdirectory in github) . I have put in docstrings as you mentioned and checked out pep8 - defintely quite useful. I am in the process of putting in tests and wanted to check in with you mainly about this. I know we did some error raising and exception related functions in the Whirlwind Tour of python and I have been researching using the unittest module - is the latter the format you would like the testing to be done in? if so, I understand that we are to have created specific testing modules each with tests for the particular module being referred to.
In general, any guidance that can be provided regarding the testing espectations for the project will be appreciated.

barlehmann commented:
Also, Circli testing is also something I do not feel yet very knowledgeable about, I plan to do some research but if you have any tips/resources you feel would be particularly recommended on this as well, I'll be glad to hear.

leej3 commented:
> using the unittest module

You will want to use pytest. They will be similar but I have set circleci up to use pytest:
https://circleci.com/gh/biof309/project_spring_2020/150?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

> doing some research on a way to create the subdirectory in github

Locally, create your subdirectory, move the appropriate files into it, git add the new files (and the deleted files), git commit, and git push. 

> I understand that we are to have created specific testing modules each with tests for the particular module being referred to

Roughly yes. The bar will be set low. I just want people to have a taste for how much easier life can be if they write tests, pass the tests, rinse and repeat. It's hard to get to that stage, and I don't expect thorough testing or anything like that. The process will hopefully give people a feeling for trying to spot semantic errors in their "working" code.

> Also, Circli testing is also something I do not feel yet very knowledgeable about, I plan to do some research but if you have any tips/resources you feel would be particularly recommended on this as well, I'll be glad to hear.

You don't really have to worry about this. It is a way for us to quickly tell if your code installs and passes some tests. The circleci documentation is a useful resource for this but you can ignore it all together if you wish. If, from your project directory, you can successfully execute the following you should be good to go:

```
pip install .
pytest
```

barlehmann committed:
going through this final uncommit due to problems pushing or pullingMerge branch 'master' of https://github.com/barlehmann/EEG_GSR_Wellness_Tracker

barlehmann committed:
changing subdirectory of project files

barlehmann committed:
adding excel sheet to new subdirectory

barlehmann committed:
adding excel sheet to new subdirectory

barlehmann commented:
Thank you very much for all of the help. I have now got my files into a subdirectory as you had specified. After watching a tutorial video on pytest, I spend about two hours today trying to get the pytest to work without getting my simple test to pass. I created a module with the correct name, and specified test functions with the correct names. Pytest seems to be running without a problem, but my tests continue to fail even though they should not be failing. I have tried changing the subfolder the test was in to see if that was causing some problems, but that did not change things either. I know these coding aspects can take some time to get familiar with, but any insight or resources you recommend will be appreciated.

barlehmann committed:
adding test files

barlehmann commented:
Also just to add some more context to the challenges in pytest - one error I continue to come up against is the following: 
../../../../../opt/anaconda3/lib/python3.7/site-packages/_pytest/python.py:501: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
../../../../../opt/anaconda3/lib/python3.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:967: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:677: in _load_unlocked
    ???
../../../../../opt/anaconda3/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:142: in exec_module
    exec(co, module.__dict__)
test_append_table.py:1: in <module>
    import append_table
E     File "/Users/barlehmann/Desktop/Python_project/github_project/EEG_GSR_Wellness_Tracker/eeg_gsr_wellness_tracker/append_table.py", line 45
E       '''
E       This extract_mean_amplitude_to_data function's  purpose is to read an EEG file (using the read_raw_eeg_file such as:
E       'AusEC.edf' is used and to extract from it overall amplitude data for (a) specific electrode(s) of interest (in this
E       case Pz is the 10-20 electrode location of interest) and to insert this information along with the timestamp of the
E       recording as a new row in the excel datafile 'Mood_Focus_Table' tabulating biometric data alongside subjective report
E       of mood and focus likert-scale ratings.
E       The extract_mean_amplitude_to_data takes the argument Picks which specifies the specific electrode(s) being looked at
E       An example of using this function might look like:
E       read_raw_eeg_file('/Users/barlehmann/desktop/AusEC.edf')
E       extract_mean_amplitude_to_data(Picks = 'EEG Fp2-LE')
E       Of course, the read_raw_eeg_file must specify a particular file to be extracting the mean amplitude from, which is why
E       this function is being used before the extraction function.
E       '''
E          
E                                                                                                                           
E                                                                                                                           
E                                                                                                                          
E                                                                                                                            
E                                              
E                                                                                                                            
E                                                         
E                                                               
E                                                           
E                                                                                                                             
E                                                                  
E         ^
E   IndentationError: expected an indented block

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

barlehmann commented:
In this last error, I have looked and cannot find any indentation issues with my files, so I am confused by this error message.

barlehmann commented:
ok I have found some other errors with my pytest file, and have made some progress and will keep you posted. No need for immediate help at this moment with pytest.

barlehmann commented:
my test_append_table.py keeps failing: 
def read_raw_eeg_file():
    eeg_file_path = '/Users/barlehmann/desktop/AusEC.edf'
    assert eeg_file_path = str

I am not sure why this should come out to an error of incorrect syntax, any tips would be very helpful.

leej3 committed:
fix double quotes

leej3 committed:
move tests into tests subdirectory

leej3 commented:
You're doing well. I have set-up a [pull request](https://github.com/barlehmann/EEG_GSR_Wellness_Tracker/pull/1) to indicate some basic fixes you can make to the project (merge it or copy the changes, whatever is easier for you). Namely

+ Fix double quotes
+ Move test modules into the appropriate directory and add imports to them.

With those changes, you should be able to pip install the project and run pytest to get sensible outputs/errors that you can begin to debug.

I don't see it in the codebase but regarding your code snippet I see a few problems:

```
def read_raw_eeg_file():
eeg_file_path = '/Users/barlehmann/desktop/AusEC.edf'
assert eeg_file_path = str
```


Should probably look more along the lines of:

```
def read_raw_eeg_file(eeg_file_path):
    assert isinstance(eeg_file_path,  str)
    ...
```

barlehmann committed:
Merge pull request #1 from leej3/make_test_and_installation_fixes

Make test and installation fixes

barlehmann committed:
saving new files

barlehmann committed:
iadding fileserge branch 'master' of https://github.com/barlehmann/EEG_GSR_Wellness_Tracker
d

barlehmann commented:
Dr. Lee, thank you for your response, this was certainly helpful to get my tests to run. However, after merging your changes, I began getting the following (below) error message after running pytest, suggesting that pytest is not connecting the graph module with the test_graph in the tests folder. I will continue to research this problem but wanted to keep you posted. 

ImportError while importing test module '/Users/barlehmann/Desktop/Python_project/github_project/EEG_GSR_Wellness_Tracker/tests/test_graph.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_graph.py:2: in <module>
    import graph
E   ModuleNotFoundError: No module named 'graph'

leej3 commented:
correct. The correct pattern when importing from the installed package is:

```
from eeg_gsr_wellness_tracker import graph
```

The way you were importing only works if you are importing files from the directory your code is running in (not convenient/robust).

leej3 commented:
Apologies. There was a __init__.py missing from that pull request. You need to include that in the directory eeg_gsr_wellness_tracker

barlehmann committed:
adding init file

barlehmann committed:
updating tests

barlehmann commented:
Dr. Lee, thanks for the information and suggestions. My tests presently both have the format you reccomended ( i.e from eeg_gsr_wellness_tracker import graph ) and I also added and empty init file to the directory you specified, however I am still getting the same error message.

barlehmann committed:
updating test file

barlehmann committed:
dddd
 Please enter the commit message for your changes. Lines starting

barlehmann committed:
d
anch master

barlehmann commented:
![image](https://user-images.githubusercontent.com/60762465/80928934-30a50580-8d76-11ea-82de-01f91609ef55.png)

barlehmann commented:
the tail end of my last message was cut off: 

![image](https://user-images.githubusercontent.com/60762465/80928954-5205f180-8d76-11ea-8aa9-cdae4e6cd1b7.png)

leej3 commented:
Thanks for the feedback. You are going about these problems in the correct way. I wouldn't be worried about what you have to show at the moment. You have demonstrated hard work and have met most of the rubric. Engaging with testing and working out some of the sordid details (which we can of course help you with) is an immensely valuable lesson to work through.

Some details to help you pass your tests both locally and on circleci:

you need to include pyopenxl and mne as dependencies for the project (install_requires argument to setup or use a requirements.txt file).

When you install the code it gets relocated into the site-packages directory of your current python interpreter. This means that  relative paths to data will both fail.   Data should not be stored in the code directory and should never be referenced using an absolute path (they will fail for everyone but you on your current computer). If you do need data it for tests (which you likely will, and in this case do), put it in the tests directory. That way, the relative path makes sense in the context of the location of the test file. 

If testing the equality of two arrays the best thing to use is probably `numpy.allclose` or a similar function. Perhaps I mislead you with that template code I gave you. I presented that to indicate the sorts of things you should be testing (input and output of functions rather than their internals).

barlehmann committed:
making all modules with relative vs absolute paths to needed files

barlehmann committed:
copying files used by modules into tests folder to ensure that tests have no problems accessing these files

barlehmann committed:
adding init file to tests directory to ensure that is able to run pytests smoothly

barlehmann commented:
![image](https://user-images.githubusercontent.com/60762465/81025885-3833df80-8e46-11ea-83af-7ce60d44b837.png)

barlehmann committed:
adding changes to test file

leej3 committed:
move data to its own directory

leej3 committed:
fix test forread_raw_eeg_file

fix paths to sample data used for tests

remove some superfluous code

add test to check that when sample data is read the first
three values are as expected

leej3 committed:
fix read_raw_eeg_file

data should not be hardcoded or stored in the code directory. These should be
generic functions that operate on paths defined by the user of the code

superfluous data is removed

global variable was removed. a good heuristic is that if you are using a
global variable you are doing something wrong

removed additional keyword arguments to the io function from mne. You only
specify keyword arguments if you want to. You don't have to specify a value
for them if you do not want.

do not use "pass". Not sure why it was included here

leej3 commented:
ok I have created a new pull request with a demonstration of a test that executes one of your functions and checks that the result is as expected. 

I have modified the function in the process. Your function was setting a global variable. It shouldn't do that. And append_table.py should not need data. That should be provided by the user.

Please read through the code/comments/and commit messages and try to understand the logic of both the test and the function it tests as I have implemented it.

barlehmann committed:
Merge pull request #2 from leej3/add_example_of_working_test_and_code

Have a look at this example as well as the commit messages

barlehmann commented:
Thank so much, that certainly helps me understand some of the issues I was having with there needing to be a data directory separate from the others, and that the global variables were causing some issues. I will certainly be looking into the logic of the test functions as you have implemented them. Much appreciated.

leej3 commented:
Great.  Glad to hear it helped. I only fixed one. I can help if you get stuck fixing the rest


> On May 5, 2020, at 18:41, barlehmann <notifications@github.com> wrote:
> 
> Thank so much, that certainly helps me understand some of the issues I was having with there needing to be a data directory separate from the others, and that the global variables were causing some issues. I will certainly be looking into the logic of the test functions as you have implemented them. Much appreciated.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

barlehmann committed:
adding changes to readme based on development of project through last month

barlehmann committed:
Update README.md

leej3 commented:
Impressive work. You really threw yourself at this. Well done. You have used of a wide range of advanced packages to achieve your goals and written it in a modular way that will make parts of it reusable for future sustainable expansion.


For future improvement, look at some of the other projects for examples of testing. For example, [this example](https://github.com/avapapetti/project_spring_2020/blob/master/tests/test_core_functions.py) of using  pandas tools for testing when using data frames will be helpful to the sort of work you are doing.

Sometimes the python style you use is odd, although often harmless from the perspective of execution of the program. I'm not sure where you picked them up. For example the syntax

```
print\
("hello")
```
Or using multiline strings in the code as a mechanism to comment. Typically text editors will nicely wrap comments and prepend them with "#" so I just use that. Overall just keep an eye on pep8 as your python skills develop.


Variable scoping was one issue that I think you should do a little reading about. Justified usage of global variables is exceptionally rare. You should typically only be using variables that you have imported from somewhere, defined in the scope of the current module, have recieved as input to a function, or defined within the current function's scope. Getting this right early on is important, otherwise you will adapt your coding in abnormal ways (like global variable usage) and suffer as a programmer in the long run.
